### ISSUE TITLE
[Feature Request] Add project context as reference

### ISSUE THREAD

**Issue by NickAcPT** [waf/CSharpRepl#10]
Hello, first of all, I hope this issue isn't too confusing.

Sometimes I'm working on C# projects where I'd like to use this C# repl on, but that would mean that I'd have to manually reference the right shared framework, nuget dependencies and other project references.

To fix that, I'm basically proposing adding a console option where one or multiple csproj files could be passed and CSharpRepl would resolve the references and everything necessary.

Since MSBuild is a pain, I recommend using https://github.com/daveaglick/Buildalyzer as it lets you get all the required information for this.

**Comment by waf** [comment]
Thanks! Agreed, this is a good feature to add.

The foundation is already there; if you use `#r` to load your application's entry assembly, it should pick up all the requested assemblies, both project and nuget assemblies, if those assemblies are present in the bin directory. This means that as long as the Shared Framework option is configured correctly, it should take a single `#r` command to get up and running.

It's still a bit clunky, though, and requires configuring the shared frameworks via command line parameters; it'd be better to allow loading projects directly in the REPL.

Some thoughts on user experience:

- We could potentially reuse the `#r` statement for this; if the referenced file is a csproj file, we could parse it (manually, or via buildalyzer) and set up everything for the user. Need to research how other REPLs handle this, if they do at all.
- For the command line configuration, do we add a new `--project` parameter, or overload the `--reference` parameter (ugly? But consistent if we use `#r`)?

**Comment by waf** [comment]
I think I've found a good solution, but it's slightly different then the issue title. @NickAcPT, would the following solution work for you?

CSharpRepl can use the autogenerated `*.runtimeconfig.json` file that's placed next to DLLs to automatically load the correct shared framework. This means, if you have a project named `MyApp.Web` (for example), to load that project in the REPL you'd run:

```csharp
#r "C:\Path\To\MyApp.Web.dll"
```

The REPL would then be able to ensure that all nuget references, assembly references, and shared frameworks are set correctly, and it doesn't require any new commands or configuration options for the end user.

A benefit of this is that we don't need to solve for multiple target frameworks or Debug/Release configuration options -- this information is embedded in the filepath of the `#r` statement.

**Comment by NickAcPT** [comment]
@waf I think that will work for me, yes. Seems to be a good compromise between both options.
👍

**Comment by waf** [comment]
@NickAcPT I've pushed branch https://github.com/waf/CSharpRepl/tree/support-referencing-csproj-sln-files with initial support for the following:

- When loading an assembly, any required shared frameworks are referenced as well (as in my previous comment).
- You can now pass both `.csproj` and `.sln` files as references (e.g. `#r "path/to/my/proj.csproj"`). csharprepl will build the project/sln and add the resulting assembly as a reference, which will take advantage of the above shared framework loading.

If you'd like, you can try the above branch and make sure it works for your projects. It comes with the following caveats:

- csproj/sln files are only supported in the `#r` statement for now, I haven't implemented support in the command line parameters (yet).
- At this point I've only tested with net5.0 assemblies, but it should work with earlier .NET Core / .NET Standard as well.

**Comment by waf** [comment]
Merged to `main`, pending release.

**Comment by waf** [comment]
This is now released in version 0.2.6 . I can't promise it'll handle every possible csproj that is thrown at it, but it works for some non-trivial projects I've tested it in, including asp.net core projects.

Update via: 

```
dotnet tool update -g csharprepl
```
